src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,143 @@
+        }
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> current = getNodeByIndex(index);
+            Node<T> prev = current.prev;
+            Node<T> newNode = new Node<>(prev, value, current);
+            current.prev = newNode;
+            if (prev == null) {
+                head = newNode;
+            } else {
+                prev.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Objects.checkIndex(index, size);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Objects.checkIndex(index, size);
+        Node<T> node = getNodeByIndex(index);
+        T prev = node.value;
+        node.value = value;
+        return prev;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Objects.checkIndex(index, size);
+        T removedElement;
+        if (index == 0) {
+            removedElement = head.value;
+            head = head.next;
+            if (head == null) {
+                tail = null;
+            }
+        } else {
+            Node<T> prev = getNodeByIndex(index - 1);
+            removedElement = prev.next.value;
+            prev.next = prev.next.next;
+            if (index == size - 1) {
+                tail = prev;
+            }
+        }
+        size--;
+        return removedElement;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(currentNode.value, object)) {
+                if (currentNode.prev == null && currentNode.next == null) {
+                    head = null;
+                    tail = null;
+                } else if (currentNode.prev == null) {
+                    head = currentNode.next;
+                    head.prev = null;
+                } else if (currentNode.next == null) {
+                    tail = currentNode.prev;
+                    currentNode.prev.next = null;
+                } else {
+                    currentNode.prev.next = currentNode.next;
+                    currentNode.next.prev = currentNode.prev;
+                }
+                size--;
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return head == null;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
```

Comment:
you might split this method in two, as suggested in common mistakes file
so you don`t iterate your nodes one way